anitya/tests/test_flask_api_v2.py

```diff
@@ -0,0 +1,212 @@
+                "error": "invalid_token",
+            }
+        self.assertEqual(data, exp)
+
+
+    def test_project_monitoring_request(self):
+        self.maxDiff = None
+        output = self.app.post('/api/v2/projects/')
+        self.assertEqual(output.status_code, 401)
+        self._check_authentication_failure_response(output)
+
+
+class _AuthenticatedAPItestsMixin(_APItestsMixin):
+    """Common test definitions for tests of authenticated access"""
+
+    def _post_app_url(self, post_url):
+        return self.app.post(post_url)
+
+    def test_invalid_project_monitoring_request(self):
+        output = self._post_app_url('/api/v2/projects/')
+        self.assertEqual(output.status_code, 400)
+        # Error details should report the missing required fields
+        data = _read_json(output)
+        error_details = data["message"]
+        self.assertIn("backend", error_details)
+        self.assertIn("homepage", error_details)
+        self.assertIn("name", error_details)
+
+
+class MockAuthenticationTests(_AuthenticatedAPItestsMixin, Modeltests):
+    """Test authenticated behaviour with authentication checks mocked out"""
+    def setUp(self):
+        super(MockAuthenticationTests, self).setUp()
+        # Replace anitya.authentication._validate_api_token
+        def _bypass_token_validation(validated, raw_api, *args, **kwds):
+            return raw_api(*args, **kwds)
+        mock_auth = mock.patch('anitya.authentication._validate_api_token',
+                               _bypass_token_validation)
+        mock_auth.start()
+        self.addCleanup(mock_auth.stop)
+
+
+_this_dir = os.path.dirname(__file__)
+CREDENTIALS_FILE = os.path.join(_this_dir, "oidc_credentials.json")
+
+
+class LiveAuthenticationTests(_AuthenticatedAPItestsMixin, Modeltests):
+    """Test authenticated behaviour with live FAS credentials"""
+
+    def setUp(self):
+        super(LiveAuthenticationTests, self).setUp()
+        if self.oidc is None:
+            self.skipTest("OpenID Connect is not configured")
+        if not os.path.exists(CREDENTIALS_FILE):
+            self.skipTest("No saved OIDC credentials available")
+        self.access_token = self._refresh_access_token()
+
+    def _post_app_url(self, post_url):
+        query_string = "access_token={0}".format(self.access_token)
+        return self.app.post(post_url, query_string=query_string)
+
+    def _refresh_access_token(self):
+        with open(CREDENTIALS_FILE) as f:
+            oidc_credentials = json.load(f)
+        client_details = oidc_credentials["client_details"]
+        client_id = client_details["client_id"]
+        client_secret = client_details["client_secret"]
+        refresh_uri = client_details["token_uri"]
+
+        token = oidc_credentials["client_token"]
+        token["expire_in"] = -1
+        oauth = requests_oauthlib.OAuth2Session(client_id, token=token)
+        token = oauth.refresh_token(refresh_uri,
```

Comment:
It might be good to catch any other exceptions here.
It is not guaranteed you'll be able to refresh your tokens with the scopes you've requested.